src/PromoService.kt

```
@@ -0,0 +1,1342 @@
+
+        return userPromos[0]
+    }
+
+    /**
+     * This method check if the product type is valid for users and promo type
+     */
+    fun isProductTypeValid(productType: ProductType, user: User, type: PromoType, promo: Promo? = null): Boolean {
+        val isProductTypeValidForUser = when (productType) {
+            ProductType.TRADE -> user.isActiveTrading && (type == PromoType.REFERRAL || promo!!.ruleEngine?.validForTradePromotion() ?: false)
+            ProductType.WEALTH -> user.isWealth && (type == PromoType.REFERRAL || promo!!.ruleEngine?.validForWealthPromotion() ?: true)
+            else -> throw IllegalStateException()
```

detekt

[warning] A call to the default constructor of an exception was detected. Instead one of the constructor overloads should be called. This allows to provide more meaningful exceptions.

(detekt.exceptions.ThrowingExceptionsWithoutMessageOrCause)